### PR TITLE
Python 3.7 - 3.10 support

### PR DESCRIPTION
**Pull request by alexcouper**
Addressing the issues in https://github.com/pretenders/pretenders/issues/145

Main issues were:
 - nose no longer being supported
 - smtpd no longer being supported
 
Additional includes getting the github actions to work in some initial form against this project.

This merge request builds for all currently supported versions of python according to https://endoflife.date/python.

3.7 through to 3.10.

### CODE REVIEW

File: pretenders/server/mock_servers/smtp/server.py [hunk 1]
```diff
@@ -8,28 +8,34 @@
...
     def run(self):
         "Runner for the MockSMTPServer"
-        LOGGER.info("Running SMTP mock server")
-        asyncore.loop()
-
+        LOGGER.info(f"Running SMTP mock server on {self.host} {self.port}")
+        controller = Controller(handler=MockSMTPHandler(self), hostname=self.host, port=self.port)
+        controller.start()
+        # TODO: Pass in our own event loop here to run this correctly
+        # https://aiosmtpd.readthedocs.io/en/latest/controller.html#aiosmtpd.controller.Controller
```
Comment (alexcouper) [inline pretenders/server/mock_servers/smtp/server.py:35]: https://aiosmtpd.readthedocs.io/en/latest/migrating.html

File: pretenders/server/mock_servers/smtp/tests/unit/test_smtp_handler.py [hunk 2]
```diff
@@ -3,7 +3,7 @@
 from mock import Mock
 
 from pretenders.common.constants import PRETEND_PORT_RANGE
-from pretenders.smtp.handler import SmtpHandler
+from pretenders.server.mock_servers.smtp.handler import SmtpHandler
```
Comment (alexcouper) [inline pretenders/server/mock_servers/smtp/tests/unit/test_smtp_handler.py:6]: Not sure how this was working before tbh

**Comment by alexcouper** [comment]
@txels this is ready for review

**Review by txels — APPROVED** [review]
Looks good. Surprisingly few changes to migrate out of smtpd. Well done.
I've added some comments re: how we may want to make CI work with less duplication, but they are not blockers, can be done later.

File: setup.py [hunk 3]
```diff
@@ -16,7 +16,7 @@
     url='https://github.com/pretenders/pretenders',
     packages=find_packages(),
     install_requires=[
-        'bottle', 'argparse'
+        'bottle', 'argparse', 'aiosmtpd'
```
Comment (txels) [inline setup.py:19]: BTW shouldn't we add something like `python_requires='>=3.7.0',`?

Comment (alexcouper) [reply setup.py:19]: Good call

File: .github/workflows/docker-image.yml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    branches: [ "develop" ]
+
+jobs:
+
+  test37:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v3
+    - name: Run tests on python 3.7
+      run: docker build . --file Dockerfile.testing.py3-7 --tag pretenders-test-py37:testing && docker run pretenders-test-py37:testing
```
[the remark this answers is not in the document]
Comment (alexcouper) [reply .github/workflows/docker-image.yml]: Not going to attempt this for now...

**Comment by txels** [comment]
Looks good to me!

**Pull request merged**